python/hail/dataset.py

```diff
@@ -4772,12 +4770,101 @@ def skat(self, variant_keys, single_key, weight_expr, y, x, covariates=[], logis
                       left_aligned=bool)
-    def split_multi(self, propagate_gq=False, keep_star_alleles=False, left_aligned=False):
+    def split_multi_generic(self, variant_expr, genotype_expr, keep_star_alleles=False, left_aligned=False):
         """Split multiallelic variants.
 
+        **Example**
+
+        :py:meth:`~hail.VariantDataset.split_multi`, which splits
+        multiallelic variants for the HTS genotype schema and updates
+        the genotype annotations by downcoding the genotype, is
+        implemented as:
+
+        >>> vds.split_multi_generic(
+        ...   'va.aIndex = aIndex, va.wasSplit = wasSplit',
+        ...   'let
+        ...         newgt = downcode(Call(g.GT), aIndex) and
+        ...         newad = if (isDefined(g.AD))
+        ...             let sum = g.AD.sum() and adi = g.AD[aIndex] in [sum - adi, adi]
+        ...           else
+        ...             NA: Array[Int] and
+        ...         newpl = if (isDefined(g.PL))
+        ...             range(3).map(i => range(g.PL.length).filter(j => downcode(Call(j), aIndex) == Call(i)).map(j => g.PL[j]).min())
+        ...           else
+        ...             NA: Array[Int] and
+        ...         newgq = gqFromPL(newpl)
+        ...    in { GT: newgt, AD: newad, DP: g.DP, GQ: newgq, PL: newPL }')
+
+        **Notes**
+
+        ``variant_expr`` is an annotation expression to update the
+        variant annotations for each output variant.  It updates `va`
+        and is evaluated in the following namespace:
+
+        - ``global``: global annotations
+        - ``v``: :ref:`variant(GR)`: the old variant
+        - ``newV``: :ref:`variant(GR)`: the new, split variant
+        - ``va``: the old variant annotations
+        - ``aIndex``: the index of the input allele to the output variant
+        - ``wasSplit``: True if the original variant was multiallelic
+
+        ``genotype_expr`` is an annotation expression to update the
+        genotype annotations for each output variant.  It updates `g`
+        and is evaluated in the following namespace:
+
+        - ``global``: global annotations
+        - ``v``: :ref:`variant(GR)`: the old variant
+        - ``newV``: :ref:`variant(GR)`: the new, split variant
+        - ``va``: the old variant annotations
+        - ``aIndex``: the index of the input allele to the output variant
+        - ``wasSplit``: True if the original variant was multiallelic
+        - ``s``: the sample
+        - ``sa``: the sample annotations
+        - ``g``: the old genotype annotations
+
+        Note, the variant and genotype annotations are updated for all
+        output variants, even those that were originally biallelic.
+        This is because the updates can change the variant annotation
+        and genotype schemas, which must be consistent across
+        variants.
+
+        :param str variant_expr: Annotation expressions to update the variant annotations for the new, split variants.
```

Comment:
Should this and `genotype_expr` be allowed to take lists?

Reply:
I think they can take multiple exprs, comma-delimited. The interface will probably look different in 0.2, so it's not worth making it marginally nicer by taking lists right now!